examples/platform/openiotsdk/app/openiotsdk_startup_gcc.cpp

```diff
@@ -0,0 +1,279 @@
+        };
+
+        osThreadId_t main_thread_id = osThreadNew(main_thread, NULL, &main_thread_attr);
+        if (main_thread_id == NULL)
+        {
+            printf("Main thread creation failed\r\n");
+            abort();
+        }
+    }
+
+    ret = osKernelStart();
+    // Note osKernelStart should never return
+    printf("Kernel failed to start: %d\r\n", ret);
+    abort();
+}
+
+/**
+ * Main thread
+ * - Initialize TF-M
+ * - Initialize the toolchain:
+ *  - Setup mutexes for malloc and environment
+ *  - Construct global objects
+ * - Run the main
+ */
+static void main_thread(void * argument)
+{
+    // Create Malloc mutex
+    {
+        osMutexAttr_t malloc_mutex_attr = { .name      = "malloc_mutex",
+                                            .attr_bits = osMutexRecursive | osMutexPrioInherit,
+                                            .cb_mem    = &malloc_mutex_obj,
+                                            .cb_size   = sizeof(malloc_mutex_obj) };
+
+        malloc_mutex = osMutexNew(&malloc_mutex_attr);
+        if (malloc_mutex == NULL)
+        {
+            printf("Failed to initialize malloc mutex\r\n");
+            abort();
+        }
+    }
+
+#ifdef TFM_SUPPORT
+    {
+        int ret = tfm_ns_interface_init();
+        if (ret != 0)
+        {
+            printf("TF-M initialization failed: %d\r\n", ret);
+            abort();
+        }
+    }
+#endif
+
+    /* Run the C++ global object constructors */
+    __libc_init_array();
+
+    int return_code = main();
+
+    exit(return_code);
+}
+
+/*
+ * Override of lock/unlock functions for malloc.
+ */
+extern "C" void __wrap___malloc_lock(struct _reent * reent)
+{
+    osMutexAcquire(malloc_mutex, osWaitForever);
+}
+
+extern "C" void __wrap___malloc_unlock(struct _reent * reent)
+{
+    osMutexRelease(malloc_mutex);
+}
+
+/*
+ * Override of new/delete operators.
+ * The override add a trace when a non-throwing new fails.
+ */
+
+void * operator new(std::size_t count)
+{
+    void * buffer = malloc(count);
+    if (!buffer)
+    {
+        printf("operator new failure from %p\r\n", CALLER_ADDR());
+        abort();
+    }
+    return buffer;
+}
+
+void * operator new[](std::size_t count)
+{
+    void * buffer = malloc(count);
+    if (!buffer)
+    {
+        printf("operator new[] failure from %p\r\n", CALLER_ADDR());
+        abort();
+    }
+    return buffer;
+}
+
+void * operator new(std::size_t count, const std::nothrow_t & tag)
+{
+    return malloc(count);
+}
+
+void * operator new[](std::size_t count, const std::nothrow_t & tag)
+{
+    return malloc(count);
+}
+
+void operator delete(void * ptr)
+{
+    free(ptr);
+}
+
+void operator delete(void * ptr, std::size_t)
+{
+    free(ptr);
+}
+
+void operator delete[](void * ptr)
+{
+    free(ptr);
+}
+
+void operator delete[](void * ptr, std::size_t)
+{
+    free(ptr);
+}
+
+/*
+ * Override of _sbrk
+ * It prints an error when the system runs out of memory in the heap segment.
+ */
+
+#undef errno
```

Comment:
Why do we need this?

Reply:
This is required to write to errno with newlib. 

See: https://www.embecosm.com/appnotes/ean9/html/ch05s03s01.html